### DISCUSSION TITLE
`${opt:}` and SLS v3 issue - are we supposed to drop `${opt:}`?

### DISCUSSION THREAD

**Discussion by lkolchin** [serverless/serverless#10954] [General]
Looking through the doco:
This - https://www.serverless.com/framework/docs/providers/aws/guide/variables tells me that `${opt:stage, 'dev'}` is resolvable

But this - https://www.serverless.com/framework/docs/guides/upgrading-v3 tells me that `${opt:foo}` variable must be replaced with `${param:foo}`

So, is `--stage` a special case and is the only one that can be resolved with `${opt:stage}` or not, because this still works for me:
```
provider:
  stage: ${opt:stage, env:STAGE_NAME, 'lab' }
```
In the rest of doco it still uses `${opt:region, self:provider.region, "us-east-1"}.` - so what's the deal with `${opt:}` and SLS v3?????

2) https://www.serverless.com/framework/docs/guides/upgrading-v3 tells me that:
all CLI options must now be passed at the end of the commands:

```
# Will no longer work in v3:
serverless --verbose deploy
 
# Correct syntax:
serverless deploy --verbose
```

But a lot of people including me use `scripts` in package.json to shorten cli commands like this:
```
  "scripts": {
    "deploy-lambda": "node_modules/.bin/serverless deploy --verbose --force",
    "remove-lambda": "node_modules/.bin/serverless remove --verbose"
  },
```

How am I supposed to use a neat if in the new syntax I have to add --verbose --force at then end?:
```
npm run deploy-lambda -- --stage dev
```

Meaning that I'd need to remove `--verbose --force` and use the following ugly syntax?
```
npm run deploy-lambda -- --stage dev --param="SlsDeploymentBucket=bucketname-${acc_id}-${region}-master" --verbose --force
```

**Comment by pgrzesik** [comment]
Hey @lkolchin,

`${opt:` still works and is supported in v3, however, it is now not possible to pass freeform cli options - `stage` or `region` are supported and recognized by the serverless commands so they're available - if you try to pass e.g. `--foo="123"` as cli option it will fail.

As for 2nd question - it's only about not supporting CLI options passed before the command (see the example) - in your case, there shouldn't be a problem and you should be able to use the same approach you've been using so far.

**Reply by lkolchin** [reply]
Thanks @pgrzesik ,

1) Yeah I've tested `${opt:` with `stage` and `region` and it worked.
It would have been nice if the doco specifically stated that only `stage` and `region` are supported with `${opt:` (or are there other supported parameters?)
but other parameters are not (for example: `--SlsDeploymentBucket "${SlsDeploymentBucket}"`)

The problem is that we don't know that 'coz it's not in the doco.

2) Thanks for clarifying that - tested and indeed it's supported.
You are saying that `it's only about not supporting CLI options passed before the command (see the example)` but you forgot to post an example?

**Reply by pgrzesik** [reply]
Hey,

1. It is explicitly mentioned in the V3 docs upgrade but maybe, not sure if it makes sense to mention it anywhere else explicitly - if you're coming on V3 as a new user, we don't promote using freeform CLI variables
2. I was referring to the example you've included in your original post from the v3 upgrade guide - it shows that `--verbose` flag cannot be put before the `deploy` command

**Reply by lkolchin** [reply]
Thanks @pgrzesik ,

Upgrade doco doesn't say what parameters still support `${opt:` - That's what I want to know!
It should be in the upgrade doco or in any searchable doco for v3........

I understand that the new way with free form params is to go with `--param=` but that's not what I was saying.

Oh, ok so it's about my example :)

**Reply by pgrzesik** [reply]
I'm impossible to list all params that are supported by `opt` because plugin commands can implement support for new options - the best way for your specific case would probably be to check `sls deploy --help` and list of supported options. Honestly, most of the options affect processing in a way that doesn't require them to be explicitly referenced in config via opt - what is the specific use case that you want to address?
